velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -435,40 +428,120 @@ RowVectorPtr NestedLoopJoinProbe::getNextCrossProductBatch(
+    return genCrossProductSingleBuildRow(
+        buildVector, outputType, probeProjections, buildProjections);
+  } else if (isCrossJoin() && isSingleBuildVector()) {
+    return genCrossProductSingleBuildVector(
+        buildVector, outputType, probeProjections, buildProjections);
+  } else {
+    return genCrossProductMultipleBuildVectors(
+        buildVector, outputType, probeProjections, buildProjections);
+  }
+}
 
-    // Project columns from the probe side.
-    projectChildren(
-        projectedChildren, input_, probeProjections, numOutputRows, nullptr);
+RowVectorPtr NestedLoopJoinProbe::genCrossProductSingleBuildRow(
+    const RowVectorPtr& buildVector,
+    const RowTypePtr& outputType,
+    const std::vector<IdentityProjection>& probeProjections,
+    const std::vector<IdentityProjection>& buildProjections) {
+  VELOX_CHECK(isSingleBuildRow());
 
-    // Wrap projections from the build side as constants.
-    for (const auto [inputChannel, outputChannel] : buildProjections) {
-      projectedChildren[outputChannel] = BaseVector::wrapInConstant(
-          numOutputRows, 0, buildVector->childAt(inputChannel));
-    }
+  std::vector<VectorPtr> projectedChildren(outputType->size());
+  size_t numOutputRows = input_->size();
+  probeRowCount_ = input_->size();
+
+  // Project columns from the probe side.
+  projectChildren(
+      projectedChildren, input_, probeProjections, numOutputRows, nullptr);
+
+  // Wrap projections from the build side as constants.
+  for (const auto [inputChannel, outputChannel] : buildProjections) {
+    projectedChildren[outputChannel] = BaseVector::wrapInConstant(
+        numOutputRows, 0, buildVector->childAt(inputChannel));
+  }
+  return std::make_shared<RowVector>(
+      pool(), outputType, nullptr, numOutputRows, std::move(projectedChildren));
+}
+
+RowVectorPtr NestedLoopJoinProbe::genCrossProductSingleBuildVector(
+    const RowVectorPtr& buildVector,
+    const RowTypePtr& outputType,
+    const std::vector<IdentityProjection>& probeProjections,
+    const std::vector<IdentityProjection>& buildProjections) {
+  VELOX_CHECK(isSingleBuildVector());
+  std::vector<VectorPtr> projectedChildren(outputType->size());
+
+  // Calculate how many probe rows we can cover without exceeding
+  // outputBatchSize_.
+  if (buildRowCount_ > outputBatchSize_) {
+    probeRowCount_ = 1;
   } else {
-    numOutputRows = buildVector->size();
-
-    // Project columns from the build side.
-    projectChildren(
-        projectedChildren,
-        buildVector,
-        buildProjections,
-        numOutputRows,
-        nullptr);
-
-    // Wrap projections from the probe side as constants.
-    for (const auto [inputChannel, outputChannel] : probeProjections) {
-      projectedChildren[outputChannel] = BaseVector::wrapInConstant(
-          numOutputRows, probeRow_, input_->childAt(inputChannel));
-    }
+    probeRowCount_ = std::min(
+        (vector_size_t)outputBatchSize_ / buildRowCount_,
```

Comment:
I think we can remove `buildRowCount_` field, all the usage can be replaced with `buildVector->size()` or `buildVectors_[0]->size()`